pkg/backends/promremotewriter/promremotewriter.go

```diff
@@ -0,0 +1,381 @@
+				fl.addMetricf(timer.Mean, timer.Tags, "%s.mean", key)
+			}
+			if !prw.disabledSubtypes.Median {
+				fl.addMetricf(timer.Median, timer.Tags, "%s.median", key)
+			}
+			if !prw.disabledSubtypes.StdDev {
+				fl.addMetricf(timer.StdDev, timer.Tags, "%s.std", key)
+			}
+			if !prw.disabledSubtypes.Sum {
+				fl.addMetricf(timer.Sum, timer.Tags, "%s.sum", key)
+			}
+			if !prw.disabledSubtypes.SumSquares {
+				fl.addMetricf(timer.SumSquares, timer.Tags, "%s.sum_squares", key)
+			}
+			for _, pct := range timer.Percentiles {
+				fl.addMetricf(pct.Float, timer.Tags, "%s.%s", key, pct.Str)
+			}
+		}
+		fl.maybeFlush()
+	})
+
+	metrics.Gauges.Each(func(key, tagsKey string, g gostatsd.Gauge) {
+		fl.addMetric(g.Value, g.Tags, key)
+		fl.maybeFlush()
+	})
+
+	metrics.Sets.Each(func(key, tagsKey string, set gostatsd.Set) {
+		fl.addMetric(float64(len(set.Values)), set.Tags, key)
+		fl.maybeFlush()
+	})
+
+	fl.finish()
+}
+
+func (prw *Client) postMetrics(ctx context.Context, buffer **bytes.Buffer, writeReq *pb.PromWriteRequest) error {
+	if err := prw.post(ctx, buffer, "metrics", writeReq); err != nil {
+		return err
+	}
+	atomic.AddUint64(&prw.seriesSent, uint64(len(writeReq.Timeseries)))
+	return nil
+}
+
+// SendEvent sends an event to NOWHERE.
+func (prw *Client) SendEvent(ctx context.Context, e *gostatsd.Event) error {
+	return nil
+}
+
+// Name returns the name of the backend.
+func (prw *Client) Name() string {
+	return BackendName
+}
+
+func (prw *Client) post(ctx context.Context, buffer **bytes.Buffer, typeOfPost string, writeReq *pb.PromWriteRequest) error {
+	post, err := prw.constructPost(ctx, buffer, writeReq)
+	if err != nil {
+		atomic.AddUint64(&prw.batchesDropped, 1)
+		return err
+	}
+
+	b := backoff.NewExponentialBackOff()
+	clck := clock.FromContext(ctx)
+	b.Clock = clck
+	b.Reset()
+	b.MaxElapsedTime = prw.maxRequestElapsedTime
+	for {
+		if err = post(); err == nil {
+			atomic.AddUint64(&prw.batchesSent, 1)
+			return nil
+		}
+
+		next := b.NextBackOff()
+		if next == backoff.Stop {
+			atomic.AddUint64(&prw.batchesDropped, 1)
+			return fmt.Errorf("[%s] %v", BackendName, err)
+		}
+
+		prw.logger.WithFields(logrus.Fields{
+			"type":  typeOfPost,
+			"sleep": next,
+			"error": err,
+		}).Warn("failed to send")
+
+		timer := clck.NewTimer(next)
+		select {
+		case <-ctx.Done():
+			timer.Stop()
+			return ctx.Err()
+		case <-timer.C:
+		}
+
+		atomic.AddUint64(&prw.batchesRetried.Cur, 1)
+	}
+}
+
+func (prw *Client) constructPost(ctx context.Context, buffer **bytes.Buffer, writeReq *pb.PromWriteRequest) (func() error /*doPost*/, error) {
+	raw, err := proto.Marshal(writeReq)
+	if err != nil {
+		return nil, err
+	}
+
+	bufBorrowed := (*buffer).Bytes()
+	bufCompressed := snappy.Encode(bufBorrowed, raw)
+	if cap(bufBorrowed) < cap(bufCompressed) {
+		// it overflowed the buffer and had to allocate, so we'll keep the larger buffer for next time
+		*buffer = bytes.NewBuffer(bufCompressed)
```

Comment:
is it worth looking at https://pkg.go.dev/sync#Pool for this functionality instead of doing the memory management for go?

Reply:
`snappy` is creating/growing the byte array if needed, the best we can do is detect if it needed to grow, and re-use it.  The `bytes.Buffer` is used in the semaphore, so we'll reach a local maxima and then stop allocating more data.  There's not a large amount of byte slices allocated.

Reply:
Sounds good.